### PR TITLE
Avoiding dilly-dally in curl_socket_timer_cb

### PR DESCRIPTION
**Pull request by aravindhank11**
Signed-off-by: Aravindhan K <aravindhank11@gmail.com>

### CODE REVIEW

**Comment by aravindhank11** [comment]
We were unnecessarily having a hash lookup in the TCP path which is removed in this PR. Also, there was a 1ms delay induced in polling the socket. Fixed it as well. 

When compared to the metrics published during the early days of TENS, we were 3x down. With this changeset, we are back on the same perf.

File: te/GET_AND_RUN_DOCKER_IMAGE.py [hunk 1]
```diff
@@ -158,7 +158,7 @@ def __update_run_cmd(self):
             self.lgr.debug("Output of executing 'which netstat' out%s= err=%s" %(out,err))
             if(bool(err) or not(bool(out))):
                 sys.exit(EXIT_NO_NETSTAT_CMD)
-            (out, err) = self.__exec_cmd("netstat -ltn | awk '{print $4}' | grep '[0-9]' | cut -d ':' -f2  | uniq")
+            (out, err) = self.__exec_cmd("netstat -ltn | awk '{print $4}' | grep '[0-9]'| grep -o '[^:]*$' | sort | uniq")
```
Comment (aravindhank11) [inline te/GET_AND_RUN_DOCKER_IMAGE.py:161]: The above command figures out free ports to allocate for internal services. It had a bug while parsing services listening on IPv4 and v6. Fixing the parser

File: te_dp/src/te_agent.c [hunk 2]
```diff
@@ -193,7 +193,6 @@ void te_create_resources()
                 wprint("Could not init curl.Exiting\n");
                 goto cleanup_res;
             }
-            te_create_socket_hashTbl(TE_SOCKET_HASH_TABLE_SIZE);
```
Comment (aravindhank11) [inline te_dp/src/te_agent.c:196]: No more hash lookup in TCP path

File: te_dp/src/te_dp.c [hunk 3]
```diff
@@ -175,9 +175,6 @@ void load_random_session_data(te_session_t * session)
 void send_session_requests(te_session_t *session, unsigned int num_requests)
 {
     unsigned int iter = 0;
-    if (unlikely(is_session_config_state_stopped(session->session_cfg_p))) {
```
Comment (aravindhank11) [inline te_dp/src/te_dp.c:178]: We anyways don't support pause and resume APIs. Removing unnecessary if check, which would always fail.

File: te_dp/src/te_sock.c [hunk 4]
```diff
@@ -58,7 +58,7 @@ void te_create_socket_hashTbl(unsigned int size)
 
 void free_socket_node(uv_handle_t *socket_poll_handle)
 {
-    te_socket_node_t *socket_node = socket_poll_handle->data;
```
Comment (aravindhank11) [inline te_dp/src/te_sock.c:61]: Renaming the struct to indicate the socket_node and there by the hash is only relevant in UDP path. We could further examine and eliminate the UDP hash table also if possible.

File: te_dp/src/te_stats_collector.cpp [hunk 5]
```diff
@@ -997,8 +997,12 @@ int main(int argc, char** argv)
         exit(1);
     }
 
-    freopen("/tmp/te_stats_collector_output.log","w",stdout);
-    freopen("/tmp/te_stats_collector_error.log","w",stderr);
+    FILE* fp;
```
Comment (aravindhank11) [inline te_dp/src/te_stats_collector.cpp:1000]: Make was complaining about not handling the return of `freopen`. Fixing the warning.

File: te_dp/src/te_tcp_dp.c [hunk 6]
```diff
@@ -174,7 +174,7 @@ void dump_error_request(te_tcp_request_t *req, int fail_reason, char* header, ch
             if (req->req_type == TE_SESSION_REQ_GET) {
                 rlist = &(res_cfg->greqs[req->prof_index][req->url_index]);
             }
-            else if (req->req_type == TE_SESSION_REQ_POST) {
+            else {
```
Comment (aravindhank11) [inline te_dp/src/te_tcp_dp.c:177]: In TCP path, we only support GETs and POSTs. Altering else-if to an else

File: te_dp/src/te_tcp_dp.c [hunk 7]
```diff
@@ -713,15 +703,10 @@ int curl_socket_timer_cb(CURLM *multi, long timeout_ms, void *userp)
...
     */
     te_session_t *session = (te_session_t *)userp;
-    if (unlikely(is_session_config_state_stopped(session->session_cfg_p))) {
-       return 0;
-    }
     if (timeout_ms < 0) {
        uv_timer_stop(&session->tcp->cm_timer);
     }
     else {
-        if(timeout_ms == 0)
```
Comment (aravindhank11) [inline te_dp/src/te_tcp_dp.c:723]: Seems like this code was directly inherited from https://curl.se/libcurl/c/multi-uv.html. We are on fast path, there is no place to dilly-dally. This change induced a 1 ms delay in the receive path

File: te_dp/src/te_tcp_dp.c [hunk 8]
```diff
@@ -734,22 +719,33 @@ int curl_socket_cb(CURL *easy, curl_socket_t s, int action, void *user_p, void *
     * If it doesn't work, still leave it. A bug takes about a week to solve
     * :""((
     */
-    te_socket_node_t* socket_ptr_cb = (te_socket_node_t*)socket_p;
```
Comment (aravindhank11) [inline te_dp/src/te_tcp_dp.c:737]: The function is the crux of the changeset. Curl in itself maintains a mapping b/w socket and user-defined socket pointed. 

When a new socket is created, user can associate his custom pointer, to the socket via `curl_multi_assign`. Post this in every callback associated with this socket, curl will give the pointer to the user. This behavior obviates the need for maintaining a hash table. 

Essentially, the previous code, always inserted the node to hash table, but never did a lookup due to the detailed behavior. With this changeset, there is no need to insert as well.

**Review by vipinpr5991 — APPROVED** [review]
Looks good to me

File: te_dp/src/te_tcp_dp.c (hunk 8 above)
Comment (sudks) [reply te_dp/src/te_tcp_dp.c:737]: @aravindhank11 : Was this impacting the initial burst of sync packets not being sent out so fast ? Can you please explain on the same.

Comment (aravindhank11) [reply te_dp/src/te_tcp_dp.c:737]: No this was an improvement, I attempted to make. However, seems like my understanding was flawed. When multiple requests are to be streamed over the same connection, curl doesn't maintain the sockfd --> pointer mapping, necessitating a hash map. 

Hash map is not needed for a 1 request over 1 connection scenarios whose scenario is anyways fringe. 

To answer your qn, the drop in perf came as a direct result of in curl_socket_timer_cb function:
```
if(timeout_ms == 0)
    timeout_ms = 1
```

**Review by sudks — APPROVED** [review]

**Comment by srikawnth** [comment]
Changes looks good @aravindhank11 , However you need to take care of pause and resume. They are unique features which brings good testable value for any Network Product with Live Sessions with connections running and pausing.

**Review by srikawnth — COMMENTED** [review]
Approving! Changes looks good @aravindhank11 , However you need to take care of pause and resume. They are unique features which brings good testable value for any Network Product having Live TCP Sessions with connections running and pausing.

**Pull request merged**